Backend/Remora.Discord.Gateway/DiscordGatewayClient.cs

```diff
@@ -423,225 +348,113 @@ private bool ShouldReconnect
 
-                var dispatch = await _responderDispatch.DispatchAsync
+                break;
+            }
+            case IPayload<IHeartbeat>:
+            {
+                await EnqueueCommandAsync
                 (
-                    receiveHello.Entity,
-                    _disconnectRequestedSource.Token
+                    new Heartbeat
+                    (
+                        _heartbeatData.LastSequenceNumber == 0
+                            ? null
+                            : _heartbeatData.LastSequenceNumber
+                    ),
+                    ct
                 );
 
-                if (!dispatch.IsSuccess)
-                {
-                    return dispatch;
-                }
-
-                // Set up the send task
-                var heartbeatInterval = hello.Data.HeartbeatInterval;
-
-                _sendTask = GatewaySenderAsync(heartbeatInterval, _disconnectRequestedSource.Token);
-
-                // Attempt to connect or resume
-                var connectResult = await AttemptConnectionAsync(stopRequested);
-                if (!connectResult.IsSuccess)
-                {
-                    return connectResult;
-                }
-
-                // Now, set up the receive task and start receiving events normally
-                _receiveTask = GatewayReceiverAsync(_disconnectRequestedSource.Token);
-
-                _log.LogInformation("Connected");
-
-                _shouldReconnect = false;
-                _isSessionResumable = false;
-                _lastReceivedHeartbeatAck = 0;
-
-                _connectionStatus = GatewayConnectionStatus.Connected;
-
                 break;
             }
-            case GatewayConnectionStatus.Connected:
-            {
-                // Check the send and receive tasks for errors
-                if (_sendTask.IsCompleted)
-                {
-                    var sendResult = await _sendTask;
-                    if (!sendResult.IsSuccess)
-                    {
-                        return sendResult;
-                    }
-                }
+        }
 
-                if (_receiveTask.IsCompleted)
-                {
-                    var receiveResult = await _receiveTask;
-                    if (!receiveResult.IsSuccess)
-                    {
-                        return receiveResult;
-                    }
-                }
+        return true;
+    }
 
-                // No need to check for errors like crazy
-                await Task.Delay(TimeSpan.FromMilliseconds(100), stopRequested);
-                break;
-            }
+    /// <inheritdoc />
+    protected override async Task<Result<TimeSpan>> SendHeartbeatAsync(CancellationToken ct)
+    {
+        if (_heartbeatData.Interval == TimeSpan.Zero)
+        {
+            // We have not received OP code hello yet, so don't heartbeat
+            return TimeSpan.FromMilliseconds(50);
         }
```

Comment:
More of a nitpick/nuance thing but [according to the docs](https://discord.dev/topics/gateway#heartbeating), you could *technically* heartbeat immediately. Not that you should, but you *could*.

Reply:
I made that change because the docs say this lol:

> You may send heartbeats before this interval elapses, but you should avoid doing so unless necessary

So yeah, we could. But, likewise, I don't think we should 😛.